docs/paper/dev/misc/nms.mdx

```diff
@@ -0,0 +1,71 @@
+---
+slug: /dev/nms
+---
+
+# NMS
+
+NMS is an abbreviation for `net.minecraft.server`. It is the package that contains Mojang's classes that are used by the
+server. This is internal code which often changes between versions. For this reason, plugins that use NMS code are
+often version specific or require different code for different versions. It is often used to refer to CraftBukkit code
+as well.
+
+## Using NMS and CraftBukkit
+
+In order to use NMS and CraftBukkit code, you may either add the dependencies to your plugin or use reflection. Depending
+on the server dependencies is often the easiest way to use NMS and CraftBukkit code. However, reflection often allows
+some code to be used across multiple versions.
+
+### Depending on NMS and CraftBukkit
+
+To access Paper internals, you can use the [paperweight-userdev](https://github.com/PaperMC/paperweight-test-plugin)
+gradle plugin. This will allow you to decompile the server and access the NMS and CraftBukkit code. You can
+then write your plugin against this code. This is the recommended way to use NMS and CraftBukkit code. You can find
+out more about this in the [Paperweight Userdev](/paper/dev/userdev) section.
+
+### Reflection
+
+Reflection is a way to access code at runtime. This allows you to access code that may not be available at compile time.
+Reflection is often used to access NMS and CraftBukkit code across multiple versions. However, reflection does come
+with performance impacts if used improperly.
+
+The internal CraftBukkit code is relocated to `org.bukkit.craftbukkit.<version>`. This means that any attempts to reflect
+must include the version. For example, `org.bukkit.craftbukkit.v1_20_R1.CraftServer` is the class for the server in
+version 1.20.1. You can access these classes easily with some reflection utilities.
+
+```java
+private static final String CRAFTBUKKIT_PACKAGE = Bukkit.getServer().getClass().getPackage().getName();
+
+public static String cbClass(String clazz) {
+    return CRAFTBUKKIT_PACKAGE + "." + clazz;
+}
+
+// You can then use this method to get the CraftBukkit class:
+Class.forName(cbClass("entity.CraftBee"));
+```
+
+Minecraft's code is obfuscated. This means that the names of classes and methods are changed to make it harder to
```

Comment:
this is only partly true, since paper fully deobfuscates this, I would say something like "in order to provide compability with legacy plugins, paper is reobfuscated at runtime" or something